src/integer/z/arithmetic/add.rs

```diff
@@ -140,6 +144,51 @@ impl Add<&Zq> for &Z {
 arithmetic_trait_borrowed_to_owned!(Add, add, Z, Zq, Zq);
 arithmetic_trait_mixed_borrowed_owned!(Add, add, Z, Zq, Zq);
 
+impl Add<&PolyOverZ> for &Z {
+    type Output = PolyOverZ;
+    /// Implements the [`Add`] trait for [`Z`] and [`PolyOverZ`] values.
+    /// [`Add`] is implemented for any combination of owned and borrowed values.
+    ///
+    /// Parameters:
+    ///  - `other`: specifies the polynomial to add to `self`
+    ///
+    /// Returns the sum of both as a [`PolyOverZ`].
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer::PolyOverZ;
+    /// use qfall_math::integer::Z;
+    /// use std::str::FromStr;
+    ///
+    /// let a: Z = Z::from(42);
+    /// let b: PolyOverZ = PolyOverZ::from_str("4  5 1 2 3").unwrap();
+    ///
+    /// let c: PolyOverZ = &a + &b;
+    /// let d: PolyOverZ = a + b;
+    /// let e: PolyOverZ = &Z::from(42) + d;
+    /// let f: PolyOverZ = Z::from(42) + &e;
+    /// ```
+    fn add(self, other: &PolyOverZ) -> Self::Output {
+        let mut added = fmpz(0);
+        let mut coeffs = fmpz(0);
+        let mut out = fmpz_poly_struct {
+            coeffs: &mut coeffs,
+            alloc: 0,
+            length: 0,
+        };
+        unsafe {
+            fmpz_poly_set(&mut out, &other.poly);
+            fmpz_poly_get_coeff_fmpz(&mut added, &other.poly, 0);
+            fmpz_add(&mut added, &self.value, &added);
+            fmpz_poly_set_coeff_fmpz(&mut out, 0, &added);
+        }
+        PolyOverZ { poly: out }
+    }
```

Comment:
```suggestion
        // check if the first coefficient of the polynomial is initiated and
        // can be addressed
        if other.is_zero() {
            PolyOverZ::from(self)
        } else {
            let out = other.clone();
            unsafe {
                fmpz_add(out.poly.coeffs, &self.value, out.poly.coeffs);
            }
            out
        }
```

I think this solution is cleaner as we only have to interact with FLINT itslef minimally. Also, your `added` is never cleared which would cause a memory-leak

Reply:
poly.coeffs adresses the first fmpz in the list of coefficients, therefore we can address the first coefficient this way (if defined)

Reply:
applied